### ISSUE TITLE
Duplicate-merging script : Suggested new rule

### ISSUE THREAD

**Issue by ckjpn** [Tatoeba/tatoeba2#851]
Maybe Horus should have the following rule added.

If a higher-numbered sentence is already correct,
and a lower-numbered sentence then becomes adopted and edited, causing duplicates,
retained the existing higher-numbered sentence, rather than the newly-changed lower-numbered one.

Here is a problem caused by the way Horus operates now.

https://tatoeba.org/eng/sentences/show/39592

This problem is caused because an existing (correct) unadopted sentence got adopted then changed.
It not only caused the loss of the first correct sentence, but caused the loss of ownership by the member who was following guidelines.  

This may not be a big deal, but it might be something that should be considered.

**Comment by PaulPeer** [comment]
+1 I noticed that problem several times

**Comment by loolmeh** [comment]
After giving this some thought, I'm skeptical about the feasibility of implementing this at the moment. We'd need a way to figure out the intent of the modification otherwise it can easily be used to hijack sentences. Using the id is pretty reliable in this regard, as the id counter is hard to reset without manually dropping tables, making it a good choice for logical time.

possible implementation sketch:
- figure out a way to capture the editing event without hitting the contributions table (doable)
- figure out a way to tell if a sentence wasn't edited by the owner to fix a small typo not to alter the structure significantly (good luck)
- use the marker rows that the code of the two above requirements produce and add a small check in the merging code's prioritize() (easy)

This also sort of uncovers an attack vector, people who have reserved lower ids at some early point in time can use them to grab ownership maliciously.

Marking this as won'tfix for now
cc: @trang @jiru

**Comment by trang** [comment]
Maybe I'm missing something but I can't think of a scenario where the intent of the modification really matters here. I think we could just rely on the `modified` field rather than the `id` field, to prioritize a sentence: the sentence that has the oldest `modified` date is the one we keep (instead of the sentence that has the lowest id is the one we keep).

* I add a sentence: `I like oranges.` (id=1)
* Someone adds a sentence: `I like music.` (id=2)
* I later change my sentence into `I like music.` for whatever reason.
* Horus will keep sentence id=2 because the modified date was the oldest.

In other words, the rule would be: if two sentences are identical, and both have an owner, delete the one that was modified the most recently. This would avoid the thief of ownership via reserving older ids.

One thing that could disrupt the behavior is that the `modified` date is not updated only when the sentence text is modified but also when someone changes one of the following:
- the language of the sentence
- the ownership (adopts or unadopts the sentence)

Maybe there are other problems I didn't think of.

**Comment by ckjpn** [comment]
I suspect that the modified date is being changed when audio is added as well.
At least, when I add audio to sentences, the sentences show up at the top of this page.
https://tatoeba.org/eng/sentences/of_user/CK

Could you explain how you think a sentence could be hijacked?

Wouldn't this be what happens?

There is a bad sentence or a good sentence that someone thinks is bad.  (id=1)
There is a good sentence. (id=2)
Sentence ID 1 is modified to match Sentence ID 2.
Sentence ID 2 is maintained.

I'm not sure how that could be considered hijacking.
Sentence ID 2 was the original sentence that was written that way.
Maintaining Sentence ID 2 is still preserving the original which I think should be how the duplicate-merging script handles this.

**Comment by jiru** [comment]
> I suspect that the modified date is being changed when audio is added as well.

I confirm. Changing the correctness updates the last modification date as well.

**Comment by trang** [comment]
> Could you explain how you think a sentence could be hijacked?

It depends on what policy we want to choose when it comes to granting ownership of a sentence.

Let's say that I have added a sentence and forgot the final punctuation.
Some time later, you add the same sentence, with the final punctuation.
Then I notice that I forgot the punctuation, and correct my sentence.
Our sentences are then duplicates of each other.

Who deserves to own the sentence? You or me?
1. You could say I deserve it, because I added the sentence first. I just forgot a period, but the meaning was there, and I corrected my mistake.
2. Or you could say that you deserve to own the sentence because you're the first one who added the correct version.

I would tend to go for option 2, but option 1 can be just as valid. And if we go for option 1, we would need a real person to look at the case, and decide who is the owner. The deduplication script doesn't have the intelligence to make that decision.

I'm going to close this issue because I think the problem comes from a higher level, and trying to introduce this new rule in the deduplication script is only a workaround and has downsides.
